kratos/geometries/point_on_geometry.h

```diff
@@ -0,0 +1,378 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+
+#if !defined(KRATOS_POINT_ON_GEOMETRY_H_INCLUDED )
+#define  KRATOS_POINT_ON_GEOMETRY_H_INCLUDED
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "geometries/geometry.h"
+
+#include "quadrature_point_geometry.h"
+
+namespace Kratos
+{
+///@name Kratos Classes
+///@{
+
+/* @class PointOnGeometry
+ * @ingroup KratosCore
+ * @brief The PointOnGeometry acts as topology for points on various types
+ *        of geometries as faces or curves.
+ */
+template<class TContainerPointType, int TLocalSpaceDimension, int TWorkingSpaceDimension>
+class PointOnGeometry
+    : public Geometry<typename TContainerPointType::value_type>
+{
+public:
+    ///@}
+    ///@name Type Definitions
+    ///@{
+
+    /// Pointer definition of PointOnGeometry
+    KRATOS_CLASS_POINTER_DEFINITION( PointOnGeometry );
+
+    typedef typename TContainerPointType::value_type PointType;
+
+    typedef Geometry<typename TContainerPointType::value_type> BaseType;
+    typedef Geometry<typename TContainerPointType::value_type> GeometryType;
+    typedef typename GeometryType::Pointer GeometryPointer;
+
+    typedef GeometryData::IntegrationMethod IntegrationMethod;
+
+    typedef typename BaseType::IndexType IndexType;
+    typedef typename BaseType::SizeType SizeType;
+
+    typedef typename BaseType::PointsArrayType PointsArrayType;
+    typedef typename BaseType::CoordinatesArrayType CoordinatesArrayType;
+    typedef typename BaseType::GeometriesArrayType GeometriesArrayType;
+    typedef typename BaseType::IntegrationPointsArrayType IntegrationPointsArrayType;
+
+    static constexpr IndexType BACKGROUND_GEOMETRY_INDEX = -1;
```

Comment:
Be carefull! the IndexType is a unsigned int and assigning -1 means the max_limit<int> and may produce warning in some compilers for assigning negative value to a unsigned type.